### ISSUE TITLE
Race condition in asyncio.Server.close

### ISSUE THREAD

**Issue by bmerry** [python/cpython#109564]
# Bug report

### Bug description:

There is a race condition when closing an asyncio.Server. Accepting a connection takes several iterations of the event loop to complete, and a call to `Server.close` in the middle will lead to exceptions.

First, some demonstration code. Run both of these concurrently, running the server with `python -X dev`:

Client:
```python
#!/usr/bin/env python3

import socket

while True:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        try:
            sock.connect(("127.0.0.1", 4445))
        except IOError:
            pass
```

Server:
```python
#!/usr/bin/env python3

import asyncio

def cb(reader, writer):
    writer.close()

def _my_attach(self):
    if self._sockets is None:
        print(self._sockets, self._active_count, flush=True)
    _orig_attach(self)

async def main():
    while True:
        server = await asyncio.start_server(cb, host="127.0.0.1", port=4445)
        await asyncio.sleep(0)
        server.close()
        await server.wait_closed()

asyncio.run(main())
```

It will repeatly output errors like this:
```
Error on transport creation for incoming connection
handle_traceback: Handle created at (most recent call last):
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
  File "/usr/lib/python3.12/asyncio/base_events.py", line 651, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.12/asyncio/base_events.py", line 618, in run_forever
    self._run_once()
  File "/usr/lib/python3.12/asyncio/base_events.py", line 1943, in _run_once
    handle._run()
  File "/usr/lib/python3.12/asyncio/events.py", line 84, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.12/asyncio/selector_events.py", line 211, in _accept_connection
    self.create_task(accept)
  File "/usr/lib/python3.12/asyncio/base_events.py", line 436, in create_task
    task = tasks.Task(coro, loop=self, name=name, context=context)
protocol: <asyncio.streams.StreamReaderProtocol object at 0x7f7bb340b350>
Traceback (most recent call last):
  File "/usr/lib/python3.12/asyncio/selector_events.py", line 230, in _accept_connection2
    transport = self._make_socket_transport(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/selector_events.py", line 72, in _make_socket_transport
    return _SelectorSocketTransport(self, sock, protocol, waiter,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/selector_events.py", line 936, in __init__
    super().__init__(loop, sock, protocol, extra, server)
  File "/usr/lib/python3.12/asyncio/selector_events.py", line 800, in __init__
    self._server._attach()
  File "/usr/lib/python3.12/asyncio/base_events.py", line 294, in _attach
    assert self._sockets is not None
AssertionError
```

Additionally, stopping the server (with Ctrl-C) prints lots of ResourceWarnings about unclosed transports and sockets.

When a connection arrives on a socket (with selector_events), it goes through the following steps before reaching `Server._attach`:
1. The socket is accepted: [here](https://github.com/python/cpython/blob/74a7f5d2dacd4c05aad0e64a275dae97d18f5355/Lib/asyncio/selector_events.py#L178)
2. `_accept_connection2` is scheduled (will only run on next event loop iteration): [here](https://github.com/python/cpython/blob/74a7f5d2dacd4c05aad0e64a275dae97d18f5355/Lib/asyncio/selector_events.py#L211)
3. The transport is created: [here](https://github.com/python/cpython/blob/74a7f5d2dacd4c05aad0e64a275dae97d18f5355/Lib/asyncio/selector_events.py#L230)
4. `Server._attach` is called: [here](https://github.com/python/cpython/blob/74a7f5d2dacd4c05aad0e64a275dae97d18f5355/Lib/asyncio/selector_events.py#L800)

It's possible for `server.close()` to be executed between steps 2 and 3, in which case `Server._sockets` has already been set to None, and the assertion is triggered. The client will presumably experience this as a connection that either closes immediately (if the garbage collector closes the socket) or is unresponsive.

This particular bug might be fixable by creating the transport synchronously in `_accept_connection`. I think there may be further potential race conditions because `connection_made` is also called [asynchronously](https://github.com/python/cpython/blob/74a7f5d2dacd4c05aad0e64a275dae97d18f5355/Lib/asyncio/selector_events.py#L948), which means it is possible for it to be called after `Server.close` has already returned. That might not break anything in asyncio itself but it will cause software that does something like the following to hang in 3.12 (related to #79033 / #104344):

```python
server.close()
for (reader, writer) in my_connections:
    writer.close()
    await writer.closed()
await server.wait_closed()
```

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-126660
* gh-131009

**Comment by Yaro1** [comment]
Hi, I would like to take that issue!

**Comment by gvanrossum** [comment]
@Yaro1 Okay, first step, can you reproduce the issue locally using the `main` branch?

**Comment by Yaro1** [comment]
@gvanrossum Yes I reproduced

**Comment by gvanrossum** [comment]
Okay go ahead and work on a fix then. Come back here if you are stuck.

**Comment by Yaro1** [comment]
okay, thank you

**Comment by Yaro1** [comment]
Well, I'm not sure about creating the transport synchronously in _accept_connection.

I tried to understand what is the reason of dividing accepting socket and creating a transport to  _accept_connection and _accept_connection2.

I found commit which include creating _accept_connection2 https://github.com/python/cpython/commit/2934262fd36c35843c01b96657047625ce2e3cf6 

But there is a problem - I guess there is not enough information in commit message:

```
Issue https://github.com/python/cpython/pull/23333: Fix BaseSelectorEventLoop._accept_connection(). Close the
  transport on error. In debug mode, log errors using call_exception_handler()
```

And the link for now indicates to another issue.

Could you help me please with an understanding of purpose _accept_connection and _accept_connection2? @gvanrossum

**Comment by gvanrossum** [comment]
I'm sorry, I recommend that you try a simpler issue.

**Comment by Yaro1** [comment]
okay I will take simpler

**Comment by ordinary-jamie** [comment]
Hello @bmerry 👋 I was able to reproduce this issue on latest;

<details><summary><i>repro code</i></summary>

```py
import asyncio
import sys
import socket

async def server():
    server = await asyncio.start_server(lambda *_: None, host="127.0.0.1", port=4445)
    await asyncio.sleep(0)
    server.close()
    await server.wait_closed()


def client():

    while True:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            try:
                sock.connect(("127.0.0.1", 4445))
            except IOError:
                pass
            finally:
                sock.close()


if __name__ == '__main__':
    if len(sys.argv) > 1 and sys.argv[1] == 'server':
        asyncio.run(server())
    else:
        client()
```

</details>

As noted in this issue, the reader callback for the listening socket (`_accept_connection`) and the coroutine scheduled for the accepted connection are separate (`_accept_connection2`). This means the socket can be closed in the period between the `_accept_connection2` coroutine being scheduled and the coroutine actually running.

However, I don't believe a code change is needed to address this, since the transport will be closed on error and throw a `CancelledError` into `_accept_connection2`; for this edge-case I would considered that sufficiently handled. And, the assertions will only be logged, and that should be turned off by clearing `PYTHONASYNCIODEBUG` or with `python -O`.

> Additionally, stopping the server (with Ctrl-C) prints lots of ResourceWarnings about unclosed transports and sockets.

Yup, I see 3 separate clean-up issues:

1. **Unclosed socket**: The transport (and the accepted socket) only gets closed if the transport is successfully created. If the transport fails to create there is no code path in `BaseSelectorEventLoop._accept_connection2` to close the socket. I'm a little hesitant to add this however, as I'm not sure if this will cause a `EBADF` if the socket gets closed twice in the happy path; what do you think, @gvanrossum? Would something like `if not transport: conn.close()` be necessary?
2. **Unclosed transport on `_SelectorTransport.__del__`**: I don't believe it is necessary to handle these `unclosed transport` warnings; I think this is working as intended.
3. **Multiple calls to `Server._wakeup`**: The `asyncio.Server` has a missing `None` check for when it wants to set its waiters; this happens in this edge-case because there is an explicit `Server.close` call in addition to the transports detaching from the server, resulting in multiple calls to `Server._wakeup`; I've made PR https://github.com/python/cpython/pull/126660 to address this

**Comment by asvetlov** [comment]
> 1. Unclosed socket: The transport (and the accepted socket) only gets closed if the transport is successfully created. 

Multiple `sock.close()` calls don't raise an exception; `conn.close()` could be called explicitly in both `except (SystemExit, KeyboardInterrupt)` and `except BaseException` branches safely.

**Comment by ordinary-jamie** [comment]
> Multiple sock.close() calls don't raise an exception; conn.close() could be called explicitly in both except (SystemExit, KeyboardInterrupt) and except BaseException branches safely.

Thanks @asvetlov; should I put this change in a different PR to #126660, since the `None` check and the socket close seem unrelated to each other?

**Comment by zero-context** [comment]
[fix.patch](https://github.com/user-attachments/files/30480437/fix.patch)

Hope this helps:

# Postmortem: Race condition in `asyncio.Server.close()` (Issue #109564)

**Component:** `asyncio.selector_events.BaseSelectorEventLoop._accept_connection2`,
`asyncio.base_events.Server._attach`, CPython 3.12.3 (verified byte-identical
to upstream tag `v3.12.3` for every file this bug touches, distro patch in
one unrelated line noted below)
**Status:** Bug reproduced **naturally** (real sockets, no mocking, no
monkeypatching) — deterministic, 100% reliable. Root cause confirmed by
reading the source. Minimal fix implemented, validated against the repro
(300/300 crashes → 0/300) and against the full `test_asyncio` suite
(2,469 tests, 0 failures). Not merged upstream — this is an independent
investigation, not a claim of authorship over the eventual fix.

---

## Summary

| | |
|---|---|
| **Verified via primary source** | Issue #109564, opened 2023-09-19, still open, labeled `type-bug`/`topic-asyncio`/`stdlib`, 11 comments |
| **Attempted upstream fixes** | Two PRs exist, **neither merged**: #126660 (a narrower `None`-check in `Server._wakeup()`) was closed in favor of a bigger rewrite; #131009 (a full `_ServerState` enum across `base_events.py`/`selector_events.py`/`proactor_events.py`) is still open, "awaiting core review", stale since 2026-04-22 |
| **Underlying mechanism** | `Server._attach()` asserts `self._sockets is not None`. It's called synchronously inside `_SelectorTransport.__init__`, itself called from `_accept_connection2()` — a coroutine that only starts running at least one full event-loop iteration after it was scheduled via `create_task()`. `Server.close()` sets `self._sockets = None` immediately and synchronously. If `close()` lands in that gap, the assert fires. |
| **Natural reproduction** | Real client connects over a real loopback socket; server closes after a swept number of `asyncio.sleep(0)` yields (0–10). **300/300 crashes at exactly 3 yields, 0/300 at every other point** — a genuine, deterministic natural trigger, not a synthetic forced case. |
| **A methodological trap avoided** | The crash site's own `call_exception_handler()` call is gated behind `if self._debug:`. With the event loop's default (non-debug) settings, this exact `AssertionError` is caught and silently dropped — no log, no callback, nothing. The bug is real and reproducible either way; it's just invisible unless `loop.set_debug(True)`, which this investigation's harness explicitly enables and documents rather than assuming a false negative meant "no bug." |
| **Fix** | In `_accept_connection2`, check `server._sockets is None` before building the protocol/transport; if the server has already closed, drop the already-accepted raw connection and return, instead of attaching to a closed server. |
| **Validation** | Natural-repro sweep: 300/300 crashes pre-patch at yields=3, 0/300 post-patch across the full 0–10 sweep (3,300 trials). Full `test_asyncio` package (all 30 test files, 2,469 tests): 0 failures, same 17 skips (Windows-only) as before the patch. |

---

## 1. Verifying the target before building anything

Issue #109564 was pulled from the real GitHub API, not assumed: open,
labeled `type-bug`/`topic-asyncio`/`stdlib`, 11 comments. Both linked PRs
were checked individually via the API/web, not just trusted from the issue
page:

- **#126660** — "Add `None` check for `asyncio.Server._wakeup`" — **closed,
  not merged**. Reviewer `ZeroIntensity` noted the original repro didn't
  reproduce consistently across platforms; maintainer `1st1` pushed back on
  the whole approach — *"instead of relying on implicit notion of actual
  state by checking attributes if they are None, add an explicit
  `self._state` enum"* — and the fix reportedly introduced extra
  `ResourceWarning` spam in previously-quiet scenarios. This PR targets a
  different symptom (`_wakeup()` called twice) than the one reproduced
  below.
- **#131009** — "add asyncio.Server state machine" — **open, unmerged**,
  addresses the *same* `_attach()` assertion this investigation reproduces
  (per `ZeroIntensity`'s review comment that it "fixes the original
  reproducer... without `ResourceWarning` spam"), via a full `_ServerState`
  enum spanning three files plus new tests. Still "awaiting core review",
  stale-labeled as of 2026-04.

So: genuinely unresolved right now, not secretly fixed and just left open —
checked, not assumed.

## 2. Confirming the local environment matches upstream

This machine's installed CPython is `3.12.3` (Ubuntu package
`3.12.3-1ubuntu0.13`). A shallow clone of upstream `python/cpython` at tag
`v3.12.3` was pulled and diffed file-by-file against the installed
`/usr/lib/python3.12/asyncio/`. Only one line differs anywhere in
`asyncio/`, in `selector_events.py`, unrelated to this bug (an extra
`self._maybe_pause_protocol()` call in `_write_ready`'s buffered-write
path — a distro backport of a separate flow-control fix). Every file this
investigation touches — `base_events.py`, `selector_events.py` around
`Server`/`_accept_connection2` — is byte-identical to the upstream tag.
The fix and all validation below are done against the clean upstream
checkout, not the distro-patched copy, so `fix.patch` applies cleanly to a
real `v3.12.3` source tree.

## 3. Reading the real mechanism

From `Lib/asyncio/selector_events.py` and `base_events.py` directly:

- The listening socket's reader callback, `_accept_connection()`, runs
  `sock.accept()` synchronously, then does
  `self.create_task(self._accept_connection2(...))` and returns. `Task`
  scheduling always defers the coroutine's first step via `call_soon()` —
  it does **not** start running in the same event-loop iteration.
- When `_accept_connection2()` finally runs (at least one iteration
  later), it synchronously builds a protocol and transport —
  `_make_socket_transport()` → `_SelectorSocketTransport.__init__()` →
  `_SelectorTransport.__init__()`, which calls `self._server._attach()`
  — all before its first `await`.
- `Server._attach()` is `assert self._sockets is not None; self._active_count += 1`.
- `Server.close()` sets `self._sockets = None` immediately, synchronously,
  the moment it's called — no awaiting, no deferral.

So the vulnerable window is exactly: *after* the accept-task has been
scheduled, *before* that task's step reaches `_attach()`. Anything that
runs `server.close()` inside that window — which is at least one full
event-loop tick wide, by construction — hits the assert.

## 4. Natural reproduction — no mocking, no forcing

`scripts/natural_repro.py`: a real `asyncio.start_server()`, a real client
via `asyncio.open_connection()` over real loopback sockets. For a swept
count of `asyncio.sleep(0)` yields (0 through 10) after the client starts
connecting, the server is closed, then a few more yields are given for any
pending accept-task to actually run (or crash) before teardown. Nothing
about the accept path itself is touched, patched, or mocked.

Result, 300 trials per point:

```
yields= 0  crashes=0/300
yields= 1  crashes=0/300
yields= 2  crashes=0/300
yields= 3  crashes=300/300
yields= 4  crashes=0/300
...
yields=10  crashes=0/300
```

Confirmed by full traceback (not just the exception type) that this is
exactly the reported crash site:

```
File ".../selector_events.py", line 230, in _accept_connection2
  transport = self._make_socket_transport(...)
File ".../selector_events.py", line 940, in __init__
  ...
File ".../selector_events.py", line 804, in __init__
  self._server._attach()
File ".../base_events.py", line 296, in _attach
  assert self._sockets is not None
AssertionError
```
with the exact context message the runtime reports:
`"Error on transport creation for incoming connection"`.

This is a **deterministic natural reproduction** — real sockets end to
end, 100% reliable at the specific timing point, not a probabilistic
window like the `aiohttp#6978` investigation's RST-timing sweep, and not a
forced/synthetic case like that investigation's diagnostic. It reproduces
because this race is scheduled by the event loop's own tick structure
(`create_task`'s mandatory one-iteration deferral), not by real-network
jitter, so it doesn't need thousands of trials or luck — the exact yield
count is what matters, and it's the same every run on this machine.

### A trap worth naming: the crash is invisible by default

`_accept_connection2`'s exception handler only calls
`self.call_exception_handler(context)` **`if self._debug:`**. Outside
debug mode — the default — the `AssertionError` is caught by
`except BaseException as exc:` and then simply dropped: no log line, no
callback, nothing observable at all. The first version of this repro
script ran 3,300 trials with `loop.set_exception_handler()` installed and
reported *zero* crashes everywhere, which would have been a false "this
doesn't reproduce" conclusion. Only after rereading the exact source at
the crash site and noticing the `if self._debug:` guard, then adding
`loop.set_debug(True)`, did the real 300/300 result appear. This is
recorded here rather than silently fixed, because it's a real way this
investigation could have produced a false negative, and it's plausible
it's part of why the bug has stayed open for two years without a reliable
public repro attached.

## 5. The fix

```diff
--- a/Lib/asyncio/selector_events.py
+++ b/Lib/asyncio/selector_events.py
@@ -218,6 +218,16 @@
         protocol = None
         transport = None
         try:
+            # gh-109564: create_task() defers this coroutine's first
+            # step by at least one loop iteration, so the server can
+            # legitimately close() in the gap between "connection
+            # accepted" and "this task actually runs". Attaching to an
+            # already-closed server crashes deep inside transport
+            # construction (Server._attach()'s assert); check up front
+            # instead and drop the already-accepted connection cleanly.
+            if server is not None and server._sockets is None:
+                conn.close()
+                return
             protocol = protocol_factory()
             waiter = self.create_future()
             if sslcontext:
```

This checks the same condition the maintainer-favored `#131009` checks
(server already closed) at the same conceptual point (before attaching),
but as a narrow, local guard rather than a `_ServerState` enum spanning
three files. It's deliberately scoped to just this reproduced crash, not
the full class of related concerns raised in the issue thread (e.g. the
secondary "`connection_made()` can fire after `close()` returns" note) —
see caveats below.

## 6. Validation

- **Natural-repro sweep, pre-patch:** 300/300 crashes at yields=3, 0/300
  everywhere else (3,300 trials total).
- **Natural-repro sweep, post-patch:** 0/300 at every point 0–10 (3,300
  trials total) — the exact scenario that reliably crashed before now
  runs clean, checked against the same script, not a rewritten one.
- **Full `test_asyncio` suite**, all 30 files, run against the patched
  checkout: **2,469 tests, 0 failures**, 17 skipped (all pre-existing,
  Windows-only: `test_windows_events`, `test_windows_utils` and a handful
  of platform-gated cases inside other files) — identical skip set to an
  unpatched run. Includes `test_selector_events.py` (121 tests) and every
  other functional/unit asyncio test in the tree, run with real sockets,
  not just the narrow module the fix touches.

## 7. Honest caveats

- **This does not implement the maintainer-preferred design.** `1st1`
  explicitly asked for an explicit `_ServerState` enum instead of
  attribute-`None` checks, on grounds that defensive `None` checks "mask
  symptoms" elsewhere. This fix is exactly that kind of local `None`
  check — it closes the specific reproduced crash without addressing the
  broader "implicit state via attributes" pattern `1st1` flagged, and
  without the `RuntimeError`-instead-of-silent-drop behavior `#131009`
  provides for callers who need to know a connection was rejected because
  the server had closed. It's offered as a narrow, independently-verified
  fix for the exact reproduced crash, not a replacement for `#131009`.
- **The secondary concern from the issue is not addressed.** The reporter
  also noted `connection_made()` can fire asynchronously after
  `Server.close()` has already returned, which could hang code that
  assumes closure is complete once `close()` returns. This investigation
  did not reproduce or fix that separately; it's a related but distinct
  claim that would need its own reproduction.
- **Dropping the connection is silent.** The fix closes the
  already-accepted socket with no notification to the client beyond the
  TCP-level close (no protocol callback fires). For most servers this is
  correct (the server is shutting down anyway), but it's a behavior
  choice worth flagging, not an obviously-free one.

---
*Investigation script (in `scripts/` alongside this README):
`natural_repro.py` — the real-socket yield-count sweep used for both the
pre-patch and post-patch results above.*

**Comment by gvanrossum** [comment]
That was obviously AI-generated. It's too long to read.
